sale_pricelist_global_rule/models/sale_order.py

```diff
@@ -0,0 +1,73 @@
+from odoo import api, fields, models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    need_recompute_pricelist_global = fields.Boolean()
+    has_pricelist_global = fields.Boolean(compute="_compute_has_pricelist_global")
+
+    def _get_cummulative_quantity(self):
+        """Compute the cummulative quantity of products in the sale order.
+        :returns: dict{
+            by_template: {product.template: qty},
+            by_categ: {product.category: qty}}
+        }
+        """
+        self.ensure_one()
+        qty_data = {
+            "by_template": {},
+            "by_categ": {},
+        }
+        for line in self.order_line.filtered("product_id"):
+            qty_in_product_uom = line.product_uom_qty
+            # Final unit price is computed
+            # according to `qty` in the default `uom_id`.
+            if line.product_uom != line.product_id.uom_id:
+                qty_in_product_uom = line.product_uom._compute_quantity(
+                    qty_in_product_uom, line.product_id.uom_id
+                )
+            key_template = line.product_id.product_tmpl_id
+            key_categ = line.product_id.categ_id
+            qty_data["by_template"].setdefault(key_template, 0.0)
+            qty_data["by_template"][key_template] += qty_in_product_uom
+            qty_data["by_categ"].setdefault(key_categ, 0.0)
+            qty_data["by_categ"][key_categ] += qty_in_product_uom
+        return qty_data
+
+    @api.depends("pricelist_id")
+    def _compute_has_pricelist_global(self):
+        for sale in self:
+            if not sale.pricelist_id:
+                sale.has_pricelist_global = False
+                continue
+            qty_data = self._get_cummulative_quantity()
+            pricelist = sale.pricelist_id.with_context(
+                pricelist_global_cummulative_quantity=qty_data
+            )
+            suitable_rule = self.env["product.pricelist.item"]
+            for line in sale.order_line:
```

Comment:
Thank you for your contribution.

```suggestion
            for line in sale.order_line.filtered("product_id"):
```

If a sales order line without a product_id (section, memo, etc) is passed to _get_product_rule, it causes _compute_price_rule to return {}, which then triggers an error indicating that the product.id key is missing.



_get_product_rule in odoo/addons
https://github.com/odoo/odoo/blob/c27d978ade9bcbea056933d8fb8b5a924e983bde/addons/product/models/product_pricelist.py#L110-L111